### PR TITLE
Various fixes to jobs and cache. Fixes #10637. Fixes #10634

### LINKED ISSUE #10634: Support Google Takeout JSON Location Data
**Is your feature request related to a problem? Please describe.**

Because I use a Canon T7i, my photographs don't have embedded location data from where they were taken. Currently I use Google Takeout to download my location data (I have Location History enabled), and then I use exiftool to add the location data to my RAW files using a shell script.

It seems though, that I can no longer get the KML data from my location history, but can get the JSON formatted data.

This means more work to get my photos GeoLocated, and its a pain already.

**Describe the solution you'd like**

What I would really like, is to have DarkTable natively understand the Google Location History JSON format, and be able to use that as well as the current GPX functionality if desired.

If changes are needed, it might be a good time to also add hooks to support other GPS formats at the same time.

**Alternatives**

Many people run a GPS logger on their phone to get GPX data, but I already have my Google location tracking turned on, I would rather not have to use another application to get my location data, and use up more battery life, when Google is already handling this for me. As well, ensuring that I run another logger before I take photographs is one more thing that can go wrong.

**Additional context**

I am more than happy to work on adding this functionality.

### PR DESCRIPTION
**Pull request by LebedevRI**
Changes:
* Proper job deduplication - e.g. do not generate same thumbnail in parallel by several workers
* Some infrastructure for deadlock detection in dtpthread (only in debug build)
* True and correct mipmap cache entry lock demoting. Previously there were cases when we ask for 'r' lock and would *still* get 'w'.
* Also, in some cases, when we previously would demote the lock (by unlocking and then locking 'r'), that entry could have already got gc'd in the meantime and locking 'r' would return empty, newly allocated 'w' locked entry, so we need to workaround it.
* Some assert()'s that may help detect similar cases in the future faster...

Overall, i expect that those changes will result in somewhat faster DT (especially LT!), but i'm failing to do any performance comparisons because with git master, *without those changes*, DT sometimes fails to render all the LT thumbnails...

### CODE REVIEW

File: src/common/cache.c [hunk 1]
```diff
@@ -163,6 +164,19 @@ dt_cache_entry_t *dt_cache_get_with_caller(dt_cache_t *cache, const uint32_t key
     cache->lru = g_list_remove_link(cache->lru, entry->link);
     cache->lru = g_list_concat(cache->lru, entry->link);
     dt_pthread_mutex_unlock(&cache->lock);
+
+#ifdef _DEBUG
+    const pthread_t writer = dt_pthread_rwlock_get_writer(&entry->lock);
+    if(mode == 'w')
+    {
+      assert(writer == pthread_self());
```
Comment (houz) [inline src/common/cache.c]: better use `pthread_equal()`

File: src/common/dtpthread.h [hunk 2]
```diff
@@ -195,32 +203,47 @@ static inline int dt_pthread_cond_wait(pthread_cond_t *cond, dt_pthread_mutex_t
...
 }
 
 #define dt_pthread_rwlock_rdlock(A) dt_pthread_rwlock_rdlock_with_caller(A, __FILE__, __LINE__)
 static inline int dt_pthread_rwlock_rdlock_with_caller(dt_pthread_rwlock_t *rwlock, const char *file, int line)
 {
-  int res = pthread_rwlock_rdlock(&rwlock->lock);
-  rwlock->cnt ++;
+  const int res = pthread_rwlock_rdlock(&rwlock->lock);
+  assert(!res);
+  assert(!(res && (rwlock->writer == pthread_self())));
```
Comment (houz) [inline src/common/dtpthread.h]: `pthread_equal()` again i guess

File: src/common/dtpthread.h [hunk 3]
```diff
@@ -228,19 +251,25 @@ static inline int dt_pthread_rwlock_rdlock_with_caller(dt_pthread_rwlock_t *rwlo
 #define dt_pthread_rwlock_wrlock(A) dt_pthread_rwlock_wrlock_with_caller(A, __FILE__, __LINE__)
 static inline int dt_pthread_rwlock_wrlock_with_caller(dt_pthread_rwlock_t *rwlock, const char *file, int line)
 {
-  int res = pthread_rwlock_wrlock(&rwlock->lock);
-  rwlock->cnt ++;
+  const int res = pthread_rwlock_wrlock(&rwlock->lock);
+  assert(!res);
+  __sync_fetch_and_add(&(rwlock->cnt), 1);
```
Comment (houz) [inline src/common/dtpthread.h]: why the __sync variants? this code is inside a write lock already, isn't it?

File: src/common/mipmap_cache.c [hunk 4]
```diff
@@ -659,23 +665,42 @@ void dt_mipmap_cache_get_with_caller(
...
+      buf->cache_entry = entry = dt_cache_get(&_get_cache(cache, mip)->cache, key, mode);
+      entry->_lock_demoting = 0;
+      dsc = (struct dt_mipmap_buffer_dsc *)buf->cache_entry->data;
+    }
+
+#ifdef _DEBUG
+    const pthread_t writer = dt_pthread_rwlock_get_writer(&(buf->cache_entry->lock));
+    if(mode == 'w')
+    {
+      assert(writer == pthread_self());
```
Comment (houz) [inline src/common/mipmap_cache.c]: `pthread_equal()`

File: src/control/jobs.c [hunk 5]
```diff
@@ -58,14 +60,15 @@ typedef struct _dt_job_t
...
-    TODO: somehow compare params. maybe we have to pass the sizeof(params) when setting the params to do a
-   memcmp, or maybe even
-          allow to pass a comparator for that.
+    NOTE: maybe allow to pass a comparator for params.
  */
 static inline int dt_control_job_equal(_dt_job_t *j1, _dt_job_t *j2)
 {
+  if(!j1 || !j2) return 0;
+  if(j1->params_size != 0 && j1->params_size == j2->params_size)
+    return (memcmp(j1->params, j2->params, j1->params_size) == 0);
```
Comment (houz) [inline src/control/jobs.c]: That looks wrong. Just because the params are the same the jobs don't have to be. For example different callbacks might both just have an integer as params which can easily be the same.

File: src/control/jobs.c [hunk 6]
```diff
@@ -58,14 +60,15 @@ typedef struct _dt_job_t
...
   return (j1->execute == j2->execute && j1->state_changed_cb == j2->state_changed_cb && j1->queue == j2->queue
-          && g_strcmp0(j1->description, j2->description));
+          && (g_strcmp0(j1->description, j2->description) == 0));
```
Comment (houz) [inline src/control/jobs.c]: Thanks for catching that missing `!`. :-)

File: src/control/jobs.c [hunk 7]
```diff
@@ -363,8 +406,9 @@ int dt_control_add_job(dt_control_t *control, dt_job_queue_t queue_id, _dt_job_t
 
         *queue = g_list_delete_link(*queue, iter);
         length--;
-        dt_control_job_set_state(job, DT_JOB_STATE_DISCARDED);
-        dt_control_job_dispose(job);
+
+        job_for_disposal = job;
```
Comment (houz) [inline src/control/jobs.c]: why did you move that down? just curious.

File: src/control/jobs.c [hunk 8]
```diff
@@ -404,10 +448,15 @@ int dt_control_add_job(dt_control_t *control, dt_job_queue_t queue_id, _dt_job_t
...
 
+  // dispose of dropped job, if any
+  dt_control_job_set_state(job_for_disposal, DT_JOB_STATE_DISCARDED);
+  dt_control_job_dispose(job_for_disposal);
+
   return 0;
 }
 
 static __thread int threadid = -1;
+static __thread pthread_t pthreadid = -1;
```
Comment (houz) [inline src/control/jobs.c]: I don't think this is possible. pthread_t just happens to be an int on Linux, but that is not the case on other platforms.

File: src/common/dtpthread.h (hunk 3 above)
Comment (LebedevRI) [inline src/common/dtpthread.h]: pthread_rwlock_wrlock() can still fail with EINVAL or EDEADLK, so it is better to preventively use __sync variants here.

NOTE: our wrappers are used only when _DEBUG is defined (i.e. -DCMAKE_BUILD_TYPE=Debug ).

File: src/control/jobs.c (hunk 8 above)
Comment (houz) [inline src/control/jobs.c]: Actually, this seems unused. Just a debug leftover?

File: src/common/dtpthread.h (hunk 3 above)
Comment (houz) [inline src/common/dtpthread.h]: Ah, you are right, it's not an issue with normal builds, and debug builds should indeed be extra careful.

File: src/control/jobs.c (hunk 8 above)
Comment (LebedevRI) [inline src/control/jobs.c]: It was meant to just stay there and show up in gdb bt's, but i'm not really sure how useful that would be.
I can as well just drop it.

File: src/control/jobs.c (hunk 5 above)
Comment (LebedevRI) [inline src/control/jobs.c]: Oh right, i should basically reuse the same comparison as the previous one (the one with description), but replace g_strcmp0 description with memcmp params.

File: src/control/jobs.c (hunk 7 above)
Comment (LebedevRI) [inline src/control/jobs.c]: **job** is not in queue, so it shouldn't need locking access to it with queue_mutex.
Since job disposal will be slower than not job disposing, and there does not seem to be any reason to keep queue_mutex locked for disposing.

Since now i need to lock queue_mutex in dt_control_run_job() (to unset job from array of running jobs), this might slightly help, but i did not do any mutex performance measurements yet.

**Comment by LebedevRI** [comment]
Notes fixed.

**Comment by LebedevRI** [comment]
As @houz said in IRC, we definitely want @hanatos to look at it.

File: src/control/jobs.c (hunk 7 above)
Comment (houz) [inline src/control/jobs.c]: Fair enough.

**Comment by maethor** [comment]
As discussed with @LebedevRI on IRC, this PR did fix some huge bugs on my install (preview and histogram disappearing, brush mask not working).
